official/utils/testing/golden.py

```diff
@@ -0,0 +1,279 @@
+    if os.path.exists(data_dir):
+      shutil.rmtree(data_dir)
+    os.mkdir(data_dir)
+
+    # Serialize graph for comparison.
+    graph_bytes = graph.as_graph_def().SerializeToString()
+    expected_file = os.path.join(data_dir, "expected_graph")
+    with open(expected_file, "wb") as f:
+      f.write(graph_bytes)
+
+    with graph.as_default():
+      init = tf.global_variables_initializer()
+      saver = tf.train.Saver()
+
+    with self.test_session(graph=graph) as sess:
+      sess.run(init)
+      saver.save(sess=sess, save_path=os.path.join(data_dir, self.ckpt_prefix))
+
+      # These files are not needed for this test.
+      os.remove(os.path.join(data_dir, "checkpoint"))
+      os.remove(os.path.join(data_dir, self.ckpt_prefix + ".meta"))
+
+      # ops are evaluated even if there is no correctness function to ensure
+      # that they can be evaluated.
+      ops = [op.eval() for op in ops_to_eval]
+
+      if correctness_function is not None:
+        results = correctness_function(*ops)
+        with open(os.path.join(data_dir, "results.json"), "wt") as f:
+          json.dump(results, f)
+
+  def _evaluate_test_case(self, name, graph, ops_to_eval, correctness_function):
+    """Determine if a graph agrees with the gold standard.
+
+    Args:
+      name: String defining the run. This will be used to define folder names
+        and will be used for random seed construction.
+      graph: The graph in which the test is conducted.
+      ops_to_eval: Ops which the user wishes to be evaluated under a controlled
+        session.
+      correctness_function: This function accepts the evaluated results of
+        ops_to_eval, and returns a list of values. This list must be JSON
+        serializable; in particular it is up to the user to convert numpy
+        dtypes into builtin dtypes.
+    """
+    data_dir = os.path.join(self.data_root, name)
+
+    # Serialize graph for comparison.
+    graph_bytes = graph.as_graph_def().SerializeToString()
+    expected_file = os.path.join(data_dir, "expected_graph")
+    with open(expected_file, "rb") as f:
+      expected_graph_bytes = f.read()
+      # The serialization is non-deterministic byte-for-byte. Instead there is
+      # a utility which evaluates the semantics of the two graphs to test for
+      # equality. This has the added benefit of providing some information on
+      # what changed.
+      #   Note: The summary only show the first difference detected. It is not
+      #         an exhaustive summary of differences.
+    differences = pywrap_tensorflow.EqualGraphDefWrapper(
+        graph_bytes, expected_graph_bytes).decode("utf-8")
+
+    with graph.as_default():
+      init = tf.global_variables_initializer()
+      saver = tf.train.Saver()
+
+    with self.test_session(graph=graph) as sess:
+      sess.run(init)
+      try:
+        saver.restore(sess=sess, save_path=os.path.join(
+            data_dir, self.ckpt_prefix))
+        if differences:
+          print()
```

Comment:
What are u trying to print here?

Reply:
I don't want the warning to get lost in a stream of warnings, so I line break to make it more obvious.

Reply:
Since print() and logging.warning might go to different output stream, I will suggest you to use tf.logging.warn for this new line as well, although an extra new line does not make the error message more obvious. Usually the error log is very long, and we will not go through all of them, so add some key words to make it easy for search might be a better idea.

Reply:
I hadn't considered that they would go to different locations. I think I'll just drop it.